File: go/oauth.go
package tiktok

import (
	"context"
	"errors"
	"net/url"
	"strings"
)

// AuthorizeURL is the user-facing OAuth consent host. Note this is the
// www.tiktok.com host, distinct from the open.tiktokapis.com API host.
const AuthorizeURL = "https://www.tiktok.com/v2/auth/authorize/"

// AuthorizationURLParams configures BuildAuthorizationURL.
type AuthorizationURLParams struct {
	ClientKey   string   // TikTok app client key.
	RedirectURI string   // Callback URL registered with the app (exact match).
	Scopes      []string // OAuth scopes to request.
	State       string   // Opaque CSRF value echoed back to the callback.

	// CodeChallenge enables the PKCE flow when non-empty.
	CodeChallenge string
	// CodeChallengeMethod defaults to "S256" when CodeChallenge is set.
	CodeChallengeMethod string
	// DisableAutoAuth, when set, forces the consent screen.
	DisableAutoAuth *bool
}

// BuildAuthorizationURL assembles the URL to redirect a user to for consent.
// It performs no network call. It is a package-level function (not a *Client
// method) because the consent URL needs only the client key, not a constructed
// client.
func BuildAuthorizationURL(p AuthorizationURLParams) (string, error) {
	if p.ClientKey == "" {
		return "", errors.New("tiktok: BuildAuthorizationURL requires ClientKey")
	}
	if p.RedirectURI == "" {
		return "", errors.New("tiktok: BuildAuthorizationURL requires RedirectURI")
	}
	if len(p.Scopes) == 0 {
		return "", errors.New("tiktok: BuildAuthorizationURL requires at least one scope")
	}
	q := url.Values{}
	q.Set("client_key", p.ClientKey)
	q.Set("response_type", "code")
	q.Set("scope", strings.Join(p.Scopes, ","))
	q.Set("redirect_uri", p.RedirectURI)
	q.Set("state", p.State)
	if p.CodeChallenge != "" {
		method := p.CodeChallengeMethod
		if method == "" {
			method = "S256"
		}
		q.Set("code_challenge", p.CodeChallenge)
		q.Set("code_challenge_method", method)
	}
	if p.DisableAutoAuth != nil {
		if *p.DisableAutoAuth {
			q.Set("disable_auto_auth", "1")
		} else {
			q.Set("disable_auto_auth", "0")
		}
	}
	return AuthorizeURL + "?" + q.Encode(), nil
}

// TokenResult is the decoded OAuth2 token response, returned by both
// ExchangeCode and RefreshAccessToken.
type TokenResult struct {
	AccessToken      string `json:"access_token"`
	ExpiresIn        int64  `json:"expires_in"`
	RefreshToken     string `json:"refresh_token"`
	RefreshExpiresIn int64  `json:"refresh_expires_in"`
	OpenID           string `json:"open_id"`
	Scope            string `json:"scope"`
	TokenType        string `json:"token_type"`
}

// ExchangeCodeParams carries the inputs to the authorization_code grant.
type ExchangeCodeParams struct {
	ClientKey    string // TikTok app client key.
	ClientSecret string // TikTok app client secret.
	Code         string // Authorization code delivered to the redirect URI.
	RedirectURI  string // The same callback URL used to obtain the code.
	CodeVerifier string // PKCE verifier; required when a code_challenge was used.
}

// ExchangeCode exchanges an authorization code for an access/refresh token pair.
// It uses application credentials and does not require AccessToken on the
// *Client.
func (c *Client) ExchangeCode(ctx context.Context, p ExchangeCodeParams) (*TokenResult, error) {
	if p.ClientKey == "" || p.ClientSecret == "" || p.Code == "" || p.RedirectURI == "" {
		return nil, errors.New("tiktok: ExchangeCodeParams requires ClientKey, ClientSecret, Code, and RedirectURI")
	}
	form := url.Values{}
	form.Set("client_key", p.ClientKey)
	form.Set("client_secret", p.ClientSecret)
	form.Set("code", p.Code)
	form.Set("grant_type", "authorization_code")
	form.Set("redirect_uri", p.RedirectURI)
	if p.CodeVerifier != "" {
		form.Set("code_verifier", p.CodeVerifier)
	}

	out := &TokenResult{}
	if err := c.doForm(ctx, c.baseURL+"/v2/oauth/token/", form, out); err != nil {
		return nil, err
	}
	return out, nil
}

// RefreshTokenParams carries the inputs to the TikTok OAuth refresh endpoint.
type RefreshTokenParams struct {
	ClientKey    string // TikTok app client key.
	ClientSecret string // TikTok app client secret.
	RefreshToken string // The refresh token previously issued to the user.
}

// RefreshTokenResult is the decoded OAuth2 refresh response.
//
// Deprecated: use TokenResult, which ExchangeCode and RefreshAccessToken both
// return. RefreshTokenResult is retained as an alias for backward compatibility.
type RefreshTokenResult = TokenResult

// RefreshAccessToken exchanges a refresh token for a new access/refresh pair.
// TikTok rotates the refresh token on every refresh, so callers must persist
// the returned RefreshToken for the next cycle. This call uses application
// credentials, not the user access token.
func (c *Client) RefreshAccessToken(ctx context.Context, p RefreshTokenParams) (*TokenResult, error) {
	if p.ClientKey == "" || p.ClientSecret == "" || p.RefreshToken == "" {
		return nil, errors.New("tiktok: RefreshTokenParams requires ClientKey, ClientSecret, and RefreshToken")
	}
	form := url.Values{}
	form.Set("client_key", p.ClientKey)
	form.Set("client_secret", p.ClientSecret)
	form.Set("grant_type", "refresh_token")
	form.Set("refresh_token", p.RefreshToken)

	out := &TokenResult{}
	if err := c.doForm(ctx, c.baseURL+"/v2/oauth/token/", form, out); err != nil {
		return nil, err
	}
	return out, nil
}

// RevokeAccessToken revokes a user access token, invalidating the grant.
func (c *Client) RevokeAccessToken(ctx context.Context, clientKey, clientSecret, accessToken string) error {
	if clientKey == "" || clientSecret == "" || accessToken == "" {
		return errors.New("tiktok: RevokeAccessToken requires clientKey, clientSecret, and accessToken")
	}
	form := url.Values{}
	form.Set("client_key", clientKey)
	form.Set("client_secret", clientSecret)
	form.Set("token", accessToken)

	var out struct{}
	return c.doForm(ctx, c.baseURL+"/v2/oauth/revoke/", form, &out)
}

File: tiktok/oauth.py
"""Standalone async OAuth 2.0 client for TikTok user-token management.

This is intentionally separate from :class:`~tiktok.client.TikTokClient`: the
OAuth flow authenticates with the *application* ``client_key`` /
``client_secret`` pair (not a user access token), and is used precisely to
*obtain* the first user access token — at which point no user token exists yet.

Typical flow::

    oauth = TikTokOAuth(client_key="...", client_secret="...")
    try:
        # 1. Send the user to the authorization URL.
        url = oauth.build_authorization_url(
            redirect_uri="https://app.example.com/callback",
            scopes=["user.info.basic", "video.publish"],
            state="csrf-token",
        )
        # 2. After the redirect back, exchange the code for tokens.
        tokens = await oauth.exchange_code(
            code="server_received_code",
            redirect_uri="https://app.example.com/callback",
        )
        # 3. Later, refresh the (rotating) access token.
        tokens = await oauth.refresh_access_token(tokens.refresh_token)
    finally:
        await oauth.aclose()

Reference: https://developers.tiktok.com/doc/oauth-user-access-token-management
"""

from __future__ import annotations

import asyncio
from types import TracebackType
from typing import Any, Self
from urllib.parse import urlencode

import aiohttp

from tiktok.config import (
    DEFAULT_TIMEOUT,
    TIKTOK_API_BASE_URL,
    TIKTOK_AUTHORIZE_URL,
)
from tiktok.exceptions import TikTokConfigError, build_api_error
from tiktok.models.oauth import TokenResponse


class TikTokOAuth:
    """Async helper for the TikTok OAuth 2.0 user-token lifecycle.

    Parameters
    ----------
    client_key:
        TikTok application client key (also called ``client_id`` in some docs).
    client_secret:
        TikTok application client secret.
    timeout:
        Total HTTP request timeout in seconds (default: ``DEFAULT_TIMEOUT``).
    """

    def __init__(
        self,
        client_key: str,
        client_secret: str,
        *,
        timeout: float = DEFAULT_TIMEOUT,
    ) -> None:
        if not client_key:
            raise TikTokConfigError("client_key must not be empty.")
        if not client_secret:
            raise TikTokConfigError("client_secret must not be empty.")

        self._client_key = client_key
        self._client_secret = client_secret
        self._timeout = aiohttp.ClientTimeout(total=timeout)
        self._session: aiohttp.ClientSession | None = None
        self._lock = asyncio.Lock()

    # ------------------------------------------------------------------
    # Context-manager support
    # ------------------------------------------------------------------

    async def __aenter__(self) -> Self:
        return self

    async def __aexit__(
        self,
        exc_type: type[BaseException] | None,
        exc_val: BaseException | None,
        exc_tb: TracebackType | None,
    ) -> None:
        await self.aclose()

    # ------------------------------------------------------------------
    # Session lifecycle
    # ------------------------------------------------------------------

    async def _get_session(self) -> aiohttp.ClientSession:
        async with self._lock:
            if self._session is None or self._session.closed:
                connector = aiohttp.TCPConnector(limit=100, ttl_dns_cache=300, ssl=True)
                self._session = aiohttp.ClientSession(connector=connector, timeout=self._timeout)
        return self._session

    async def aclose(self) -> None:
        """Close the underlying HTTP session and release all connections."""
        async with self._lock:
            if self._session is not None and not self._session.closed:
                await self._session.close()
                self._session = None

    # ------------------------------------------------------------------
    # Authorization URL (no network call)
    # ------------------------------------------------------------------

    def build_authorization_url(
        self,
        *,
        redirect_uri: str,
        scopes: list[str],
        state: str,
        code_challenge: str | None = None,
        code_challenge_method: str = "S256",
        disable_auto_auth: bool | None = None,
    ) -> str:
        """Construct the URL to redirect a user to for consent.

        This performs no network call; it simply assembles the query string.

        Parameters
        ----------
        redirect_uri:
            Callback URL registered with your TikTok app.  Must match exactly.
        scopes:
            OAuth scopes to request (e.g. ``["user.info.basic", "video.publish"]``).
        state:
            Opaque value echoed back to the callback; use it for CSRF protection.
        code_challenge:
            PKCE code challenge.  Required for desktop/PKCE flows; omit for the
            standard server-side ``authorization_code`` flow.
        code_challenge_method:
            PKCE challenge method (default ``"S256"``).  Ignored when
            ``code_challenge`` is ``None``.
        disable_auto_auth:
            When ``True``, forces the consent screen even if the user has
            previously authorised the app.

        Returns:
            The fully-qualified authorization URL.
        """
        params: dict[str, str] = {
            "client_key": self._client_key,
            "response_type": "code",
            "scope": ",".join(scopes),
            "redirect_uri": redirect_uri,
            "state": state,
        }
        if code_challenge is not None:
            params["code_challenge"] = code_challenge
            params["code_challenge_method"] = code_challenge_method
        if disable_auto_auth is not None:
            params["disable_auto_auth"] = "1" if disable_auto_auth else "0"
        return f"{TIKTOK_AUTHORIZE_URL}?{urlencode(params)}"

    # ------------------------------------------------------------------
    # Token grants
    # ------------------------------------------------------------------

    async def exchange_code(
        self,
        *,
        code: str,
        redirect_uri: str,
        code_verifier: str | None = None,
    ) -> TokenResponse:
        """Exchange an authorization ``code`` for an access/refresh token pair.

        Parameters
        ----------
        code:
            The authorization code delivered to your ``redirect_uri``.
        redirect_uri:
            The same callback URL used to obtain ``code``.
        code_verifier:
            PKCE code verifier.  Required when the authorization URL was built
            with a ``code_challenge``.

        Returns:
            :class:`~tiktok.models.oauth.TokenResponse`.

        Raises:
            :class:`~tiktok.exceptions.TikTokAPIError`: The grant was rejected.
        """
        form: dict[str, str] = {
            "client_key": self._client_key,
            "client_secret": self._client_secret,
            "code": code,
            "grant_type": "authorization_code",
            "redirect_uri": redirect_uri,
        }
        if code_verifier is not None:
            form["code_verifier"] = code_verifier
        payload = await self._post_token(form)
        return TokenResponse.model_validate(payload)

    async def refresh_access_token(self, refresh_token: str) -> TokenResponse:
        """Exchange a refresh token for a fresh access/refresh token pair.

        TikTok rotates the refresh token on every refresh, so callers must
        persist the returned :attr:`~tiktok.models.oauth.TokenResponse.refresh_token`
        for the next cycle.

        Parameters
        ----------
        refresh_token:
            The refresh token previously issued to the user.

        Returns:
            :class:`~tiktok.models.oauth.TokenResponse` with the new tokens.

        Raises:
            :class:`~tiktok.exceptions.TikTokAPIError`: The refresh was rejected.
        """
        if not refresh_token:
            raise TikTokConfigError("refresh_token must not be empty.")
        form = {
            "client_key": self._client_key,
            "client_secret": self._client_secret,
            "grant_type": "refresh_token",
            "refresh_token": refresh_token,
        }
        payload = await self._post_token(form)
        return TokenResponse.model_validate(payload)

    async def revoke_access_token(self, access_token: str) -> None:
        """Revoke a user access token, invalidating the grant.

        Parameters
        ----------
        access_token:
            The user access token to revoke.

        Raises:
            :class:`~tiktok.exceptions.TikTokAPIError`: The revocation failed.
        """
        if not access_token:
            raise TikTokConfigError("access_token must not be empty.")
        form = {
            "client_key": self._client_key,
            "client_secret": self._client_secret,
            "token": access_token,
        }
        await self._post_form(f"{TIKTOK_API_BASE_URL}/v2/oauth/revoke/", form)

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    async def _post_token(self, form: dict[str, str]) -> dict[str, Any]:
        """POST the token endpoint and return the parsed flat JSON body."""
        return await self._post_form(f"{TIKTOK_API_BASE_URL}/v2/oauth/token/", form)

    async def _post_form(self, url: str, form: dict[str, str]) -> dict[str, Any]:
        """POST a ``x-www-form-urlencoded`` body and surface OAuth2 errors."""
        session = await self._get_session()
        async with session.post(
            url,
            data=form,
            headers={"Content-Type": "application/x-www-form-urlencoded"},
        ) as response:
            payload: dict[str, Any] = await response.json(content_type=None)
            has_oauth_error = bool(payload.get("error"))
            if response.status >= 400 or has_oauth_error:
                raise build_api_error(
                    code=str(payload.get("error", "oauth_error")),
                    message=str(payload.get("error_description", "")),
                    log_id=str(payload.get("log_id", "")),
                    http_status=response.status,
                )
            return payload
